src/fluree/db/json_ld/transact.cljc

```diff
@@ -564,3 +564,130 @@
    (go-try
      (let [staged (<? (stage-ledger ledger fuel-tracker json-ld opts))]
        (<? (ledger-proto/-commit! ledger staged))))))
+
+;; ----------------------------------------
+
+(defn next-id
+  "Generate the next subject id - `counter` is either :last-pid or :last-sid."
+  [iri->sid counter iri]
+  (let [iri->sid* (swap! iri->sid
+                         (fn [iri->sid]
+                           (if (get iri->sid iri)
+                             iri->sid
+                             (-> iri->sid
+                                 (assoc iri (get iri->sid counter))
+                                 (update counter inc)))))]
+    (get iri->sid* iri)))
+
+(defn ->tx-state2
+  [db]
+  (let [{:keys [schema branch ledger policy], db-t :t} db
+        iri->sid (atom {:last-pid (jld-ledger/last-pid db)
+                        :last-sid (jld-ledger/last-sid db)})
+        commit-t  (-> (ledger-proto/-status ledger branch) branch/latest-commit-t)
+        t         (-> commit-t inc -) ;; commit-t is always positive, need to make negative for internal indexing
+        db-before (dbproto/-rootdb db)]
+    {:db-before                db-before
+     :policy                   policy
+     :stage-update?            (= t db-t) ;; if a previously staged db is getting updated again before committed
+     :t                        t
+     :iri->sid                 iri->sid
+     :next-pid                 (partial next-id iri->sid :last-pid)
+     :next-sid                 (partial next-id iri->sid :last-sid)
+     :iris                     (volatile! {})}))
+
+(defn generate-flakes
+  [db fuel-tracker parsed-txn tx-state]
+  (go
+    (let [error-ch  (async/chan)
+          update-ch (->> (where/search db parsed-txn fuel-tracker error-ch)
+                         (update/modify2 db parsed-txn tx-state fuel-tracker error-ch)
+                         (into-flakeset fuel-tracker))]
+      (async/alt!
+        error-ch ([e] e)
+        update-ch ([flakes] flakes)))))
+
+(defn subject-mods
+  [new-flakes db-before]
+  (go-try
+    (let [has-target-objects-of-shapes (shacl/has-target-objects-of-rule? db-before)]
+      (loop [[s-flakes & r] (partition-by flake/s new-flakes)
+             subj-mods      {}]
+        (if s-flakes
+          (let [sid            (flake/s (first s-flakes))
+                new-subject?   (first (filterv #(= const/$xsd:anyURI (flake/p %)) s-flakes))
```

Comment:
minor, but using `first` negates any benefits gained from using `filterv` because `first` will first transform the vector back into a seq. `nth` is much faster than `first` on vectors.

Reply:
I tried this but unfortunately the vector can sometimes be empty and then an `(nth [] 0)` call will blow up with index-out-of-bounds. I'm mainly using the `filterv` to avoid the overhead of laziness, so I think it's still fine, but we will have to pay the price of the `first` call.

Reply:
there's no penalty for laziness if you're going to use first anyway. the penalty only comes in if you always consume the entire sequence. i think it's best to just use filterv in this case, or build the vector and check if it's empty with `seq` before calling `nth`. in any case, it's not such a big deal